File: backend/internal/sync/sync.go
// Package sync implements FlowStock's leaderless branch-to-branch replication.
//
// There is no central server. Every node can both serve sync requests (over
// the same HTTP mux as the app) and dial peers. A sync round is stateless and
// symmetric — push what the peer lacks, then pull what we lack — so any
// topology works (pair, hub-and-spoke, full mesh) and a branch that was
// offline simply catches up on its next reachable round.
//
// All endpoints require Authorization: Bearer <shared secret>. With no secret
// configured the endpoints reject every request (fail closed).
package sync

import (
	"bytes"
	"context"
	"crypto/subtle"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	"flowstock/backend/internal/store"
)

// Batch bounds how many ops travel per request.
const Batch = 2000

// Engine wires the store to the network. SecretFn is read on every request so
// rotating the secret in settings takes effect immediately.
type Engine struct {
	Store    *store.Store
	SecretFn func() string
	// FolderFn, if set, returns the shared-folder path for file-based transport
	// (empty = disabled). Read on every round so toggling it takes effect live.
	FolderFn func() string
	NodeID   string
	client   *http.Client
	mu       sync.Mutex // serializes outbound rounds
}

func New(s *store.Store, secretFn func() string) *Engine {
	return &Engine{
		Store:    s,
		SecretFn: secretFn,
		NodeID:   s.NodeID(),
		client:   &http.Client{Timeout: 20 * time.Second},
	}
}

type opsMsg struct {
	NodeID string     `json:"node_id"`
	Ops    []store.Op `json:"ops"`
}

type pullReq struct {
	Vector map[string]string `json:"vector"`
}

func (e *Engine) authed(r *http.Request) bool {
	secret := e.SecretFn()
	if secret == "" {
		return false
	}
	presented := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
	return subtle.ConstantTimeCompare([]byte(presented), []byte(secret)) == 1
}

// Handler returns the /api/sync/* routes.
func (e *Engine) Handler() http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc("GET /api/sync/ping", func(w http.ResponseWriter, r *http.Request) {
		io.WriteString(w, "flowstock")
	})

	mux.HandleFunc("GET /api/sync/vector", func(w http.ResponseWriter, r *http.Request) {
		if !e.authed(r) {
			http.Error(w, "unauthorized", http.StatusUnauthorized)
			return
		}
		vec, err := e.Store.Vector()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, map[string]any{"node_id": e.NodeID, "org_id": e.Store.OrgID(), "vector": vec})
	})

	mux.HandleFunc("POST /api/sync/ops", func(w http.ResponseWriter, r *http.Request) {
		if !e.authed(r) {
			http.Error(w, "unauthorized", http.StatusUnauthorized)
			return
		}
		var msg opsMsg
		if err := json.NewDecoder(r.Body).Decode(&msg); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		applied, err := e.Store.ApplyOps(msg.Ops)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, map[string]any{"applied": applied})
	})

	mux.HandleFunc("POST /api/sync/pull", func(w http.ResponseWriter, r *http.Request) {
		if !e.authed(r) {
			http.Error(w, "unauthorized", http.StatusUnauthorized)
			return
		}
		var req pullReq
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		ops, err := e.Store.OpsAfter(req.Vector, Batch)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, opsMsg{NodeID: e.NodeID, Ops: ops})
	})

	return mux
}

// Result reports the outcome of syncing one peer.
type Result struct {
	PeerID  string `json:"peer_id"`
	OK      bool   `json:"ok"`
	Pushed  int    `json:"pushed"`
	Pulled  int    `json:"pulled"`
	Adopted bool   `json:"adopted"`
	Error   string `json:"error"`
}

// SyncPeer runs one full round against a peer base URL.
func (e *Engine) SyncPeer(ctx context.Context, peerID, baseURL string) Result {
	e.mu.Lock()
	defer e.mu.Unlock()

	res := Result{PeerID: peerID}
	secret := e.SecretFn()
	if secret == "" {
		res.Error = "sync secret not set"
		return res
	}
	base := strings.TrimRight(baseURL, "/")
	auth := "Bearer " + secret

	// 1. Learn the peer's vector (and workspace id), then push everything it
	// lacks. The window vector advances past each pushed batch so batches never
	// overlap.
	peerVec, peerOrg, err := e.fetchVector(ctx, base, auth)
	if err != nil {
		res.Error = err.Error()
		return res
	}
	// Pairing: a brand-new node adopts the workspace it is joining. An
	// established node keeps its own workspace, and mismatched ops are rejected
	// on apply, so two real workspaces never silently merge over a shared secret.
	if adopted, aerr := e.Store.AdoptOrg(peerOrg); aerr != nil {
		res.Error = aerr.Error()
		return res
	} else if adopted {
		res.Adopted = true
	}
	if myOrg := e.Store.OrgID(); peerOrg != "" && myOrg != "" && peerOrg != myOrg {
		res.Error = fmt.Sprintf("peer is a different workspace (%s ≠ %s); refusing to sync", peerOrg, myOrg)
		return res
	}
	window := peerVec
	for {
		ops, err := e.Store.OpsAfter(window, Batch)
		if err != nil {
			res.Error = err.Error()
			return res
		}
		if len(ops) == 0 {
			break
		}
		for _, op := range ops {
			if op.HLC > window[op.NodeID] {
				window[op.NodeID] = op.HLC
			}
		}
		if err := e.postOps(ctx, base, auth, ops); err != nil {
			res.Error = err.Error()
			return res
		}
		res.Pushed += len(ops)
		if len(ops) < Batch {
			break
		}
	}

	// 2. Pull everything we lack.
	for {
		myVec, err := e.Store.Vector()
		if err != nil {
			res.Error = err.Error()
			return res
		}
		ops, err := e.pull(ctx, base, auth, myVec)
		if err != nil {
			res.Error = err.Error()
			return res
		}
		if len(ops) == 0 {
			break
		}
		n, err := e.Store.ApplyOps(ops)
		if err != nil {
			res.Error = err.Error()
			return res
		}
		res.Pulled += n
		if len(ops) < Batch {
			break
		}
	}

	res.OK = true
	return res
}

func (e *Engine) fetchVector(ctx context.Context, base, auth string) (map[string]string, string, error) {
	req, _ := http.NewRequestWithContext(ctx, "GET", base+"/api/sync/vector", nil)
	req.Header.Set("Authorization", auth)
	resp, err := e.client.Do(req)
	if err != nil {
		return nil, "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, "", fmt.Errorf("vector: HTTP %d", resp.StatusCode)
	}
	var body struct {
		Vector map[string]string `json:"vector"`
		OrgID  string            `json:"org_id"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, "", err
	}
	if body.Vector == nil {
		body.Vector = map[string]string{}
	}
	return body.Vector, body.OrgID, nil
}

func (e *Engine) postOps(ctx context.Context, base, auth string, ops []store.Op) error {
	buf, _ := json.Marshal(opsMsg{NodeID: e.NodeID, Ops: ops})
	req, _ := http.NewRequestWithContext(ctx, "POST", base+"/api/sync/ops", bytes.NewReader(buf))
	req.Header.Set("Authorization", auth)
	req.Header.Set("Content-Type", "application/json")
	resp, err := e.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return fmt.Errorf("push: HTTP %d", resp.StatusCode)
	}
	return nil
}

func (e *Engine) pull(ctx context.Context, base, auth string, vec map[string]string) ([]store.Op, error) {
	buf, _ := json.Marshal(pullReq{Vector: vec})
	req, _ := http.NewRequestWithContext(ctx, "POST", base+"/api/sync/pull", bytes.NewReader(buf))
	req.Header.Set("Authorization", auth)
	req.Header.Set("Content-Type", "application/json")
	resp, err := e.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("pull: HTTP %d", resp.StatusCode)
	}
	var msg opsMsg
	if err := json.NewDecoder(resp.Body).Decode(&msg); err != nil {
		return nil, err
	}
	return msg.Ops, nil
}

// TestPeer checks reachability + auth against a peer URL.
func (e *Engine) TestPeer(ctx context.Context, baseURL string) bool {
	base := strings.TrimRight(baseURL, "/")
	req, _ := http.NewRequestWithContext(ctx, "GET", base+"/api/sync/vector", nil)
	req.Header.Set("Authorization", "Bearer "+e.SecretFn())
	resp, err := e.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == 200
}

// SyncAll syncs every enabled peer (or one, if only != "") and records the
// outcome on the peers table.
func (e *Engine) SyncAll(ctx context.Context, only string) []Result {
	peers, err := e.Store.ListPeers()
	if err != nil {
		return nil
	}
	var results []Result
	for _, p := range peers {
		if !p.Enabled || (only != "" && p.ID != only) {
			continue
		}
		res := e.SyncPeer(ctx, p.ID, p.URL)
		status := "ok: pushed " + itoa(res.Pushed) + ", pulled " + itoa(res.Pulled)
		if !res.OK {
			status = "error: " + res.Error
		}
		e.Store.UpdatePeerStatus(p.ID, time.Now().UTC().Format(time.RFC3339), status)
		results = append(results, res)
	}
	// Flush to / drain from the shared folder too, if one is configured. This
	// runs even with no peers, so a folder-only (offline / sneakernet) topology
	// works with no network sync at all.
	if e.FolderFn != nil {
		if dir := e.FolderFn(); dir != "" {
			e.FolderSync(dir)
		}
	}
	return results
}

// RunBackground syncs all enabled peers every interval until ctx is cancelled.
func (e *Engine) RunBackground(ctx context.Context, interval time.Duration) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			e.SyncAll(ctx, "")
		}
	}
}

func writeJSON(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(v)
}

func itoa(n int) string { return fmt.Sprintf("%d", n) }

File: src-tauri/src/sync.rs
//! Branch-to-branch sync.
//!
//! Every node can expose an HTTP listener (axum) and/or dial peers. Sync is
//! stateless and pull-based on both sides: a round consists of
//!   1. `GET  /sync/vector`   → learn what the peer has seen,
//!   2. `POST /sync/ops`      → push ops the peer lacks,
//!   3. `POST /sync/pull`     → fetch ops we lack (batched).
//! Both endpoints require `Authorization: Bearer <sync secret>`; the listener
//! refuses to start without a secret (fail closed).

use crate::store::{self, Op};
use crate::AppState;
use axum::{
    extract::State,
    http::{HeaderMap, StatusCode},
    routing::{get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::oneshot;

pub const BATCH: usize = 2000;

#[derive(Serialize, Deserialize)]
pub struct PullReq {
    pub vector: HashMap<String, String>,
}

#[derive(Serialize, Deserialize)]
pub struct OpsMsg {
    pub node_id: String,
    pub ops: Vec<Op>,
}

fn authed(state: &AppState, headers: &HeaderMap) -> bool {
    let secret = state.sync_secret();
    if secret.is_empty() {
        return false;
    }
    let presented = headers
        .get("authorization")
        .and_then(|v| v.to_str().ok())
        .and_then(|v| v.strip_prefix("Bearer "))
        .unwrap_or("");
    // Constant-time-ish comparison.
    presented.len() == secret.len()
        && presented
            .bytes()
            .zip(secret.bytes())
            .fold(0u8, |acc, (a, b)| acc | (a ^ b))
            == 0
}

async fn get_vector(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<Json<serde_json::Value>, StatusCode> {
    if !authed(&state, &headers) {
        return Err(StatusCode::UNAUTHORIZED);
    }
    let conn = state.db.lock().unwrap();
    let vector = store::vector(&conn).map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    Ok(Json(json!({ "node_id": state.node_id, "vector": vector })))
}

async fn post_ops(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(msg): Json<OpsMsg>,
) -> Result<Json<serde_json::Value>, StatusCode> {
    if !authed(&state, &headers) {
        return Err(StatusCode::UNAUTHORIZED);
    }
    let applied = {
        let mut conn = state.db.lock().unwrap();
        let mut clock = state.clock.lock().unwrap();
        store::apply_ops(&mut conn, &mut clock, &msg.ops)
            .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?
    };
    state.bump_data_version();
    Ok(Json(json!({ "applied": applied })))
}

async fn post_pull(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<PullReq>,
) -> Result<Json<OpsMsg>, StatusCode> {
    if !authed(&state, &headers) {
        return Err(StatusCode::UNAUTHORIZED);
    }
    let conn = state.db.lock().unwrap();
    let ops = store::ops_after(&conn, &req.vector, BATCH)
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    Ok(Json(OpsMsg { node_id: state.node_id.clone(), ops }))
}

pub fn router(state: Arc<AppState>) -> Router {
    Router::new()
        .route("/sync/vector", get(get_vector))
        .route("/sync/ops", post(post_ops))
        .route("/sync/pull", post(post_pull))
        .route("/sync/ping", get(|| async { "flowstock" }))
        .with_state(state)
}

/// Start (or restart) the sync listener according to current settings.
/// Fails closed: no secret → no listener.
pub fn restart_listener(state: &Arc<AppState>) -> Result<Option<String>, String> {
    // Stop any previous listener.
    if let Some(stop) = state.listener_stop.lock().unwrap().take() {
        let _ = stop.send(());
    }
    let cfg = state.sync_config();
    if !cfg.listen {
        return Ok(None);
    }
    if cfg.secret.is_empty() {
        return Err("sync secret is not set — refusing to listen".into());
    }
    let addr = format!("{}:{}", cfg.bind_addr, cfg.port);
    let (tx, rx) = oneshot::channel::<()>();
    *state.listener_stop.lock().unwrap() = Some(tx);
    let app = router(state.clone());
    let addr_for_msg = addr.clone();
    state.runtime.spawn(async move {
        let listener = match tokio::net::TcpListener::bind(&addr).await {
            Ok(l) => l,
            Err(e) => {
                eprintln!("[sync] listen {addr} failed: {e}");
                return;
            }
        };
        let _ = axum::serve(listener, app)
            .with_graceful_shutdown(async {
                let _ = rx.await;
            })
            .await;
    });
    Ok(Some(addr_for_msg))
}

/// Sync every enabled peer (or one, if `only` is set) and record the outcome
/// on the peers table. Shared by the manual "Sync now" command and the
/// background loop.
pub async fn sync_all(state: Arc<AppState>, only: Option<String>) -> Vec<SyncResult> {
    let peers: Vec<(String, String)> = {
        let conn = state.db.lock().unwrap();
        let mut stmt = match conn.prepare("SELECT id, url FROM peers WHERE enabled = 1") {
            Ok(s) => s,
            Err(_) => return Vec::new(),
        };
        let rows = match stmt.query_map([], |r| Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?))) {
            Ok(r) => r,
            Err(_) => return Vec::new(),
        };
        rows.filter_map(|r| r.ok())
            .filter(|(id, _)| only.as_deref().map(|p| p == id).unwrap_or(true))
            .collect()
    };
    let mut results = Vec::new();
    for (id, url) in peers {
        let res = sync_with_peer(state.clone(), id.clone(), url).await;
        {
            let conn = state.db.lock().unwrap();
            let status = if res.ok {
                format!("ok: pushed {}, pulled {}", res.pushed, res.pulled)
            } else {
                format!("error: {}", res.error)
            };
            let _ = conn.execute(
                "UPDATE peers SET last_sync_at = ?1, last_status = ?2 WHERE id = ?3",
                rusqlite::params![chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true), status, id],
            );
        }
        results.push(res);
    }
    results
}

#[derive(Debug, Serialize)]
pub struct SyncResult {
    pub peer_id: String,
    pub ok: bool,
    pub pushed: usize,
    pub pulled: usize,
    pub error: String,
}

/// Run one full sync round against a peer URL (e.g. "http://192.168.1.20:7365").
pub async fn sync_with_peer(state: Arc<AppState>, peer_id: String, url: String) -> SyncResult {
    let mut result = SyncResult { peer_id, ok: false, pushed: 0, pulled: 0, error: String::new() };
    let secret = state.sync_secret();
    if secret.is_empty() {
        result.error = "sync secret not set".into();
        return result;
    }
    let base = url.trim_end_matches('/').to_string();
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(15))
        .build()
        .unwrap();
    let auth = format!("Bearer {secret}");

    let round = async {
        // 1. Peer's vector → push what it lacks.
        let peer_vec: serde_json::Value = client
            .get(format!("{base}/sync/vector"))
            .header("authorization", &auth)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;
        // The window vector starts at the peer's vector and advances past each
        // pushed batch, so batches never overlap.
        let mut window: HashMap<String, String> =
            serde_json::from_value(peer_vec.get("vector").cloned().unwrap_or_default())
                .unwrap_or_default();

        let mut pushed = 0usize;
        loop {
            let ops = {
                let conn = state.db.lock().unwrap();
                store::ops_after(&conn, &window, BATCH)?
            };
            if ops.is_empty() {
                break;
            }
            for op in &ops {
                let entry = window.entry(op.node_id.clone()).or_default();
                if op.hlc > *entry {
                    *entry = op.hlc.clone();
                }
            }
            let n = ops.len();
            client
                .post(format!("{base}/sync/ops"))
                .header("authorization", &auth)
                .json(&OpsMsg { node_id: state.node_id.clone(), ops })
                .send()
                .await?
                .error_for_status()?;
            pushed += n;
            if n < BATCH {
                break;
            }
        }

        // 2. Pull what we lack.
        let mut pulled = 0usize;
        loop {
            let my_vector = {
                let conn = state.db.lock().unwrap();
                store::vector(&conn)?
            };
            let msg: OpsMsg = client
                .post(format!("{base}/sync/pull"))
                .header("authorization", &auth)
                .json(&PullReq { vector: my_vector })
                .send()
                .await?
                .error_for_status()?
                .json()
                .await?;
            if msg.ops.is_empty() {
                break;
            }
            let n = msg.ops.len();
            {
                let mut conn = state.db.lock().unwrap();
                let mut clock = state.clock.lock().unwrap();
                store::apply_ops(&mut conn, &mut clock, &msg.ops)?;
            }
            pulled += n;
            if n < BATCH {
                break;
            }
        }
        Ok::<(usize, usize), Box<dyn std::error::Error + Send + Sync>>((pushed, pulled))
    };

    match round.await {
        Ok((pushed, pulled)) => {
            result.ok = true;
            result.pushed = pushed;
            result.pulled = pulled;
            if pulled > 0 {
                state.bump_data_version();
            }
        }
        Err(e) => result.error = e.to_string(),
    }
    result
}
